homeassistant/components/knx/sensor.py

```diff
@@ -28,22 +32,34 @@ async def async_setup_platform(
     entities = []
     for device in hass.data[DOMAIN].xknx.devices:
         if isinstance(device, XknxSensor):
-            entities.append(KNXSensor(device))
+            entities.append(KNXSensor(hass, device))
     async_add_entities(entities)
 
 
 class KNXSensor(KnxEntity, SensorEntity):
     """Representation of a KNX sensor."""
 
-    def __init__(self, device: XknxSensor) -> None:
-        """Initialize of a KNX sensor."""
+    def __init__(self, hass: HomeAssistantType, device: XknxSensor) -> None:
+        """Initialize a KNX sensor."""
         self._device: XknxSensor
+        if device.ha_value_template is not None:
+            device.ha_value_template.hass = hass
         super().__init__(device)
 
     @property
     def state(self) -> StateType:
         """Return the state of the sensor."""
-        return self._device.resolve_state()
+        state = self._device.resolve_state()
+        template = self._device.ha_value_template
+        if template is not None and state is not None:
+            try:
+                return template.async_render({"value": state})  # type: ignore[no-any-return]
```

Comment:
We have template sensors for this purpose. We don't mix templates into device interfacing integrations.

Reply:
Oh, I was under the impression that this is what `value_template` is for. 
Like stated in https://www.home-assistant.io/docs/configuration/templating/ and used by many other integrations (MQTT, Rest, CommandLine, etc.).
Having the users to configure and manage 2 entities seems unfortunate when the KNX-sensor is then only used as input for the template.

Reply:
That's the separation we have. There are some exceptions like you mentioned.

Reply:
🤔 so what would it take to join this circle of excellence? Maybe if we moved it from "sensor" to a new "raw" platform that results in a SensorEntity or BinarySensorEntity?
It is just not possible to cover all cases of possible values received over the KNX bus, especially when it comes to devices that roll their own DPTs. Events are fine, but don't have the convenience of requesting state updates.

Reply:
That's our current design. You can ask for a second opinion if you want.

Reply:
I've tried that for the template cover, but this is not user friendly at all.

A KNX entity is generally determined by several KNX addresses. If I now want to change only a single value I have to break up the complete KNX entity into its parts, by defining sensors/switches for the underlying KNX addresses. Afterwards I have to define the functionality of the KNX entity via template xxx completely new. By doing that I loose functionality of the KNX integration. In the cover case for instance the KNX integration offers a position computation and an opening/closing state while the cover runs. To mimic this via template cover is difficult or not even possible.

Reply:
Ok... will close this issue. 

Before I put any effort in it: would it be acceptable to use value templates for the knx expose function?
`expose` registers `async_track_state_change_event`and sends the value (state or attribute) of an existing HA entity to an address on the KNX bus - so no HA entities are created. Some values (eg enum types like climate presets) can't be converted directly or need to be transformed to comply knx standards. So it would be convenient to have a value_template forming the raw payload (list[int] mostly).

Reply:
The expose feature should not be part of the knx config at all if we would follow our architecture design. It should just be a service for users.

Reply:
@mptei You can just expose a HA value to the Knx bus and use this (dummy) GA in your knx cover/climate etc configuration. This is exactly where value_template in expose would shine. 

With 2021.4 new Trigger-based template sensors we can (kind of) use templates in sensors with knx_event anyway (but this is raw bytes - so *really* not user focused).